### DISCUSSION TITLE
Managing Priority Shifts Sequencing in OptaPlanner Rostering

### DISCUSSION THREAD

**Discussion by govindsuresh1996** [TimefoldAI/timefold-solver#203] [General]
I'm working on implementing an employee rostering system using OptaPlanner, and I've run into a scenario that I'd appreciate some guidance on.

In my system, I have two types of shifts: priority 1 shifts and priority 2 shifts. These shifts are identified using a priority field in the Shift class, which is also the planning entity.

My goal is to ensure that OptaPlanner considers filling priority 2 shifts only after all priority 1 shifts are filled. Essentially, I want to enforce a sequencing constraint where priority 1 shifts should be assigned first, and only then should the solver move on to priority 2 shifts.

Could anyone provide guidance on how I can configure OptaPlanner to respect this priority order and ensure that priority 2 shifts are only assigned after all priority 1 shifts are filled? Any suggestions, code examples, or insights into how to structure the constraint configuration would be highly appreciated.

**Comment by triceo** [comment]
A very similar question was [asked and answered on StackOverflow](https://stackoverflow.com/questions/76884674/optaplanner-employee-rostering-fill-priority-1-shifts-before-priority-2-shifts).

**Reply by govindsuresh1996** [reply]
I have tried using a hard score for unassigned P1 shifts and medium/soft score for unassigned P2 shifts, but its behaving differently for different load(number of employees/shifts). It lacks consistency in the solution. So was thinking if we could use something like a pinning filter or shadow variables. Is that possible ?

**Reply by triceo** [reply]
Please explain what you mean by "lacks consistency".

**Reply by govindsuresh1996** [reply]
Sometimes it fills all the P1 shifts and then proceed on to P2 shifts and sometimes it doesn't. Even though there are enough people to fill in all the P1 shifts.

**Reply by triceo** [reply]
As explained on StackOverflow, it does not "proceed" from P1 to P2; this line of thinking will not produce results. The solver looks for a solution that best satisfies all of the constraints. If it's not finding the solution you think it should be finding, one of the following is typically the reason:

- A [score trap](https://timefold.ai/docs/timefold-solver/latest/score-calculation/score-calculation.html#scoreTrap).
- Slow score speed and/or too little solving time. As the number of employees/shifts grows, so will the time required to get a good solution.
- Little move diversity. It will be significantly easier to escape local optima if the solver has a large selection of diverse moves to chose from. Introduce new kinds of moves, such as [pillar moves](https://timefold.ai/docs/timefold-solver/latest/move-and-neighborhood-selection/move-and-neighborhood-selection.html#pillarMoveSelectors).
- Sub-optimal configuration of local search. Use the [benchmarker](https://timefold.ai/docs/timefold-solver/latest/benchmarking-and-tweaking/benchmarking-and-tweaking.html) and compare late acceptance (the default) with other algorithms, such as tabu search or simulated annealing.

**Reply by govindsuresh1996** [reply]
Thanks, I'll focus on implementing these suggestions to enhance the effectiveness of the solver in finding optimal solutions.

But out of curiosity, is there a way other than using the constraints to enforce this condition on the application.

**Reply by triceo** [reply]
I don't think so. You're asking for a guarantee that P1 shifts will all be assigned, before P2 are even being considered. That guarantee does not exist, not even if you apply all of my advice. If you need that guarantee, then a solver based on local search is not the tool you're looking for.

**Reply by nicksi** [reply]
I agree with @triceo that enforcing the priority constraint would cripple the local search solver efficiency. Although you can try to implement entity filter class that will allow only priority 1 shifts, create two local search sections in config and apply this filter in first section. So you will have all priority one shifts assigned first and on second local search run solver will try to fit priority 2 shifts into the gaps lefts by priority 1. This experiment is relatively easy to implement and will give you something to compare with alternative approaches.

Doing that do not forget to put in some termination rules for first local search. Unimproved step count or unimproved seconds limits.

**Reply by govindsuresh1996** [reply]
Thanks @nicksi, I have come up with a mock solverConfig.xml for this approach. Cud u please review it and see if I am moving in the right direction

```
<localSearch>
    <termination>
        <unimprovedStepCountLimit>100</unimprovedStepCountLimit>
        <!-- Other termination conditions -->
    </termination>
    <selector>
        <entitySelector id="priority1ShiftSelector">
            <filterClass>com.example.Priority1ShiftFilter</filterClass>
        </entitySelector>
        <!-- Other selectors -->
    </selector>
    <!-- Other configuration options -->
</localSearch>

<localSearch>
    <termination>
        <unimprovedStepCountLimit>100</unimprovedStepCountLimit>
        <!-- Other termination conditions -->
    </termination>
    <selector>
        <!-- No entity filter, includes both priority 1 and priority 2 shifts -->
        <!-- Other selectors -->
    </selector>
    <!-- Other configuration options -->
</localSearch>
```

Also, Is there a way to separate or multiple pinning filters for each phase?

**Reply by nicksi** [reply]
Yes, i've got my idea. Be ready to experiment with termination count values, for second section you can put something bigger like 500 or 1000. You could also benefit from using different search strategies for different sections. Something like late acceptance or great deludge for first local search and tabu for the second one.

**Reply by govindsuresh1996** [reply]
Is there a way to add separate or multiple pinning filters for each phase?

**Reply by nicksi** [reply]
none AFAIK - pinning filter has no idea at what phase solver is. Entity filter / values filter / selection filter/ custom move are your tools if you want to alter solver behavior between phases

**Reply by govindsuresh1996** [reply]
Where can I find more about these tools, is there a document or an example that u cud share?

**Reply by nicksi** [reply]
https://timefold.ai/docs/timefold-solver/latest/move-and-neighborhood-selection/move-and-neighborhood-selection.html#filteredSelection - filters of all types

https://timefold.ai/docs/timefold-solver/latest/move-and-neighborhood-selection/move-and-neighborhood-selection.html#customMoves -  custom moves. In that particular case custom move is not needed imho.

**Reply by govindsuresh1996** [reply]
Thanks @nicksi , appreciate your assistance and the time you took to point me in the right direction.
